packages/gatsby/src/utils/source-nodes.ts

```diff
@@ -77,12 +81,13 @@ function getStaleNodes(state: IGatsbyState, nodes: Array<Node>): Array<Node> {
 /**
  * Find all stale nodes and delete them
  */
-function deleteStaleNodes(state: IGatsbyState, nodes: Array<Node>): void {
+function deleteStaleNodes(
+  state: IGatsbyState,
+  nodes: GatsbyIterable<IGatsbyNode>
+): void {
   const staleNodes = getStaleNodes(state, nodes)
 
-  if (staleNodes.length > 0) {
-    staleNodes.forEach(node => store.dispatch(deleteNode(node)))
-  }
+  staleNodes.forEach(node => store.dispatch(deleteNode(node)))
```

Comment:
there is no `.length`on `GatsbyIterable`, but `.forEach` on iterables with 0 elements is largely no-op anyway